proto/pulumi/provider.proto

```diff
@@ -363,81 +373,135 @@ message CheckFailure {
+    // old and new states.
+    bool inputDiff = 2;
 
+    // The type of property diff kinds.
     enum Kind {
-        ADD = 0;            // this property was added
-        ADD_REPLACE = 1;    // this property was added, and this change requires a replace
-        DELETE = 2;         // this property was removed
-        DELETE_REPLACE = 3; // this property was removed, and this change requires a replace
-        UPDATE = 4;         // this property's value was changed
-        UPDATE_REPLACE = 5; // this property's value was changed, and this change requires a replace
+        // This property was added.
+        ADD = 0;
+
+        // This property was added, and this change requires a replace.
+        ADD_REPLACE = 1;
+
+        // This property was removed.
+        DELETE = 2;
+
+        // This property was removed, and this change requires a replace.
+        DELETE_REPLACE = 3;
+
+        // This property's value was changed.
+        UPDATE = 4;
+
+        // This property's value was changed, and this change requires a replace.
+        UPDATE_REPLACE = 5;
     }
 }
 
+// `DiffResponse` is the type of responses sent by a [](pulumirpc.ResourceProvider.DiffConfig) or
+// [](pulumirpc.ResourceProvider.Diff) call. A `DiffResponse` indicates whether a resource is unchanged, requires
+// updating (that is, can be changed "in place"), or requires replacement (that is, must be destroyed and recreated
+// anew). Legacy implementations may also signal that it is unknown whether there are changes or not.
+//
+// `DiffResponse` has evolved since its inception and there are now a number of ways that providers can signal their
+// intent to callers:
+//
+// * *Simple diffs* utilise the `changes` field to signal which fields are responsible for a change, and the `replaces`
+//   field to further communicate which changes (if any) require a replacement as opposed to an update.
+//
+// * *Detailed diffs* are those with `hasDetailedDiff` set, and utilise the `detailedDiff` field to provide a more
+//   granular view of the changes that have occurred. Detailed diffs are designed to allow providers to control
+//   precisely which field names are displayed as responsible for a change, and to signal more accurately what kind of
+//   change occurred (e.g. a field was added, deleted or updated).
 message DiffResponse {
-    repeated string replaces = 1; // if this update requires a replacement, the set of properties triggering it.
-    repeated string stables = 2;  // an optional list of properties that will not ever change.
-    bool deleteBeforeReplace = 3; // if true, this resource must be deleted before replacing it.
-    DiffChanges changes = 4;      // if true, this diff represents an actual difference and thus requires an update.
+    // A set of properties which have changed and whose changes require the resource being diffed to be replaced. The
```

Comment:
What to do if a nested property a.b.c is the one causing replacement? I *think* provider is expected to put "a" in the replaces and put a.b.c replace diff in detailed Diff but great to clarify.

Reply:
Not sure; good question. Tracking in https://github.com/pulumi/pulumi/issues/17691